i6_models/assemblies/conformer/conformer_v2.py

```diff
@@ -0,0 +1,132 @@
+    mhsa_cfg: ConformerMHSAV1Config
+    conv_cfg: ConformerConvolutionV1Config
+    swap_mhsa_conv: bool = True
+
+
+class ConformerBlockV2(nn.Module):
+    """
+    Conformer block module
+    """
+
+    def __init__(self, cfg: ConformerBlockV2Config):
+        """
+        :param cfg: conformer block configuration with subunits for the different conformer parts
+        """
+        super().__init__()
+        self.ff1 = ConformerPositionwiseFeedForwardV1(cfg=cfg.ff_cfg)
+        self.mhsa = ConformerMHSAV1(cfg=cfg.mhsa_cfg)
+        self.conv = ConformerConvolutionV1(model_cfg=cfg.conv_cfg)
+        self.ff2 = ConformerPositionwiseFeedForwardV1(cfg=cfg.ff_cfg)
+        self.final_layer_norm = torch.nn.LayerNorm(cfg.ff_cfg.input_dim)
+        self.swap_mhsa_conv = cfg.swap_mhsa_conv
+
+    def forward(self, x: torch.Tensor, /, sequence_mask: torch.Tensor) -> torch.Tensor:
+        """
+        :param x: input tensor of shape [B, T, F]
+        :param sequence_mask: mask tensor where 1 defines positions within the sequence and 0 outside, shape: [B, T]
+        :return: torch.Tensor of shape [B, T, F]
+        """
+        x = 0.5 * self.ff1(x) + x  #  [B, T, F]
+        if not self.swap_mhsa_conv:
+            x = self.mhsa(x, sequence_mask) + x  #  [B, T, F]
+            x = self.conv(x) + x  #  [B, T, F]
+        else:
+            x = self.conv(x) + x  # [B, T, F]
+            x = self.mhsa(x, sequence_mask) + x  # [B, T, F]
+        x = 0.5 * self.ff2(x) + x  #  [B, T, F]
+        x = self.final_layer_norm(x)  #  [B, T, F]
+        return x
+
+
+@dataclass
+class ConformerEncoderV2Config(ModelConfiguration):
+    """
+    Attributes:
+        num_layers: Number of conformer layers in the conformer encoder
+        loss_scales: Dictionary defining the loss layers and corresponding loss scales e.g. {6: 0.3, 12:1}
+        frontend: A pair of ConformerFrontend and corresponding config
+        block_cfg: Configuration for ConformerBlockV2
+    """
+
+    num_layers: int
+    loss_scales: Dict[int, float]
+
+    # nested configurations
+    frontend: ModuleFactoryV1
+    block_cfg: ConformerBlockV2Config
+
+
+class ConformerEncoderV2(nn.Module):
+    """
+    Modification compared to ConformerEncoderV1:
+    1. enable adding auxiliary losses with different loss scales
+        (C.f. https://arxiv.org/abs/2102.03216 for more details of auxiliary loss)
+    2. enable swapping the MHSA and Conv module
+        (C.f. https://arxiv.org/abs/2011.10798 for intuition behind swapping)
+    """
+
+    def __init__(self, cfg: ConformerEncoderV2Config):
+        """
+        :param cfg: conformer encoder configuration with subunits for frontend and conformer blocks
+        """
+        super().__init__()
+
+        self.frontend = cfg.frontend()
+        self.module_list = torch.nn.ModuleList([ConformerBlockV2(cfg.block_cfg) for _ in range(cfg.num_layers)])
+        self.output_layers = list(cfg.loss_scales.keys())
```

Comment:
I am wondering if you really need to pass the whole `loss_scales` dict since the scales are not used in the `forward function`. Instead, you can pass the list of indicies directly.

Reply:
but later in train step, we need the loss scale, I think it would be easier to get the scales from `ctc.conformer. loss_scales` instead of passing it as an argument to train step function

Reply:
I think Mohammad is right here, the `loss_scales` should be attached to your (own. custom) model config, not the encoder config. You can even put the parameter for which layers to return into the forward instead of the class That way a user can have different auxiliary losses during different training stages without having to modify the encoder code here.

Reply:
This is not resolved, I would still vote to have a `return_layers: List[int]` in the forward instead of having this static in the config. @mmz33 @albertz what is your opinion on this?

Reply:
@Judyxujj I know this might be overkill, but one use case would be to have "moving" auxiliary losses, that are not at fixed layers but start in early layers and move higher over the course of training.